shopping_mall/backend/ai/agent/executor.py

```
@@ -0,0 +1,652 @@
+        truncated.rfind("?"),
+    )
+    if last_sentence_end > MAX_ANSWER_LENGTH // 2:
+        truncated = truncated[: last_sentence_end + 1]
+
+    return truncated + "\n\n(이어지는 내용은 상담원에게 문의해 주세요.)"
+
+
+def _log_trace(trace: "list[TraceStep]", question: str) -> None:
+    """추론 과정을 INFO 레벨로 출력."""
+    if not trace:
+        logger.info(f"[trace] 질문='{question[:60]}' → 도구 호출 없음 (직접 답변)")
```

Ruff (0.15.9)

[warning] Logging statement uses f-string

(G004)